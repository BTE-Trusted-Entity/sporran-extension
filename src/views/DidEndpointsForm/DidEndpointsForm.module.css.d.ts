export const __esModule: true;
export const container: string;
export const heading: string;
export const subline: string;
export const didLine: string;
export const loading: string;
export const list: string;
export const card: string;
export const add: string;
export const cardHeading: string;
export const expand: string;
export const addCollapse: string;
export const collapsedCard: string;
export const collapsedUrl: string;
export const collapsedType: string;
export const buttons: string;
export const collapse: string;
export const remove: string;
export const labelLine: string;
export const label: string;
export const input: string;
export const buttonsLine: string;
export const submit: string;
export const cancel: string;
export const errorTooltipField: string;
export const errorTooltip: string;
export const endpoint: string;
export const name: string;
export const value: string;
