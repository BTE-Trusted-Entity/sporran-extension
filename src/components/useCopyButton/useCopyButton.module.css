.copy {
  composes: buttonIcon from '../Button/Button.module.css';
  background-image: url('../images/copy.svg');
  background-position: center 40%;
  background-size: auto 50%;
  width: 2.25rem;
  height: 2.25rem;
  position: absolute;
  right: 0.25rem;
}

.copied {
  composes: copy;
  background-image: url('../images/checkmark-circle.svg');
}
