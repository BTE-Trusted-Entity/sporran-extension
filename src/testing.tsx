export * from '@testing-library/react';
export { waitForElementToBeRemoved } from '@testing-library/dom';
import { act, render as externalRender, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import dialogPolyfill from 'dialog-polyfill';

import { AccountsProviderMock } from './testing/AccountsProviderMock';

export {
  AccountsProviderMock,
  accountsMock,
} from './testing/AccountsProviderMock';

export { mockBackgroundScript } from './testing/mockBackgroundScript';

export function render(
  ui: Parameters<typeof externalRender>[0],
  options?: Parameters<typeof externalRender>[1],
): ReturnType<typeof externalRender> {
  return externalRender(
    <AccountsProviderMock>
      <MemoryRouter>{ui}</MemoryRouter>
    </AccountsProviderMock>,
    options,
  );
}

export async function waitForTooltipUpdate(): Promise<void> {
  expect(await screen.findByRole('status')).toHaveAttribute(
    'data-popper-escaped',
  );
}

const dialogPromise = Promise.resolve();
(dialogPolyfill.registerDialog as jest.Mock).mockReturnValue(dialogPromise);

export async function waitForDialogUpdate(): Promise<void> {
  await act(() => dialogPromise);
}
