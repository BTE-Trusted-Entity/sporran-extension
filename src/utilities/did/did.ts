import {
  ConfigService,
  Did,
  DidDocument,
  DidEncryptionKey,
  DidUri,
  Utils,
} from '@kiltprotocol/sdk-js';

import { useAsyncValue } from '../useAsyncValue/useAsyncValue';

export function isFullDid(did: DidUri | undefined): boolean {
  if (!did) {
    // could be a legacy identity without DID
    return false;
  }
  return Did.parse(did).type === 'full';
}

export async function getFullDidDocument(did: DidUri): Promise<DidDocument> {
  const api = ConfigService.get('api');
  return Did.linkedInfoFromChain(await api.call.did.query(Did.toChain(did)))
    .document;
}

export function useFullDidDocument(did: DidUri): DidDocument | undefined {
  return useAsyncValue(getFullDidDocument, [did]);
}

export async function getDidDocument(did: DidUri): Promise<DidDocument> {
  return isFullDid(did)
    ? await getFullDidDocument(did)
    : Did.parseDocumentFromLightDid(did);
}

export function getDidEncryptionKey({
  keyAgreement,
}: DidDocument): DidEncryptionKey {
  if (!keyAgreement || !keyAgreement[0]) {
    throw new Error('encryptionKey is not defined somehow');
  }
  return keyAgreement[0];
}

export function parseDidUri(did: DidUri): ReturnType<typeof Did.parse> & {
  fullDid: DidUri;
} {
  const parsed = Did.parse(did);
  const fullDid = parsed.type === 'full' ? did : Did.getFullDidUri(did);

  return {
    ...parsed,
    fullDid,
  };
}

export async function needLegacyDidCrypto(
  did: DidUri | undefined,
): Promise<boolean> {
  if (!did) {
    // DID was deactivated, no action needed.
    return false;
  }
  if (!isFullDid(did)) {
    return false;
  }

  try {
    const encryptionKey = getDidEncryptionKey(await getDidDocument(did));
    return (
      Utils.Crypto.u8aToHex(encryptionKey.publicKey) ===
      '0xf2c90875e0630bd1700412341e5e9339a57d2fefdbba08de1cac8db5b4145f6e'
    );
  } catch {
    // getDidDocument might throw if the DID is not on-chain anymore (removed, another endpoint),
    // no legacy crypto needed in that case
    return false;
  }
}
