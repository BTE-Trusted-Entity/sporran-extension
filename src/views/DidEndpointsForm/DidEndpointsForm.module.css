.container {
  composes: container from '../../components/View/View.module.css';
}

.heading {
  composes: screenHeadline from '../../components/Typography/Typography.module.css';
}

.subline {
  composes: subline from '../../components/Typography/Typography.module.css';
  margin: 0 0 0.8125rem;
}

.didLine {
  composes: didLine from '../../components/CopyValue/CopyValue.module.css';
  margin-top: 1rem;
}

.loading {
  composes: spinner from '../../components/LoadingSpinner/LoadingSpinner.module.css';
  margin: 5rem auto;
}

.list {
  composes: fullWidthContainer from '../../components/View/View.module.css';
  list-style: none;
  height: calc(250rem / 16);
  margin-top: 0.75rem;
}

.card {
  composes: card from '../../components/Card/Card.module.css';
  padding-inline: 1.5rem;
  padding-bottom: 1rem;
  text-align: left;
  position: relative;
}

.card:last-child {
  height: 18rem;
}

.add {
  composes: card;
}

.cardHeading {
  composes: buttonText from '../../components/Typography/Typography.module.css';
  font-size: 0.75rem;
  padding: 0.75rem 0;
  margin: 0;
}

.expand {
  composes: expand from '../../components/Card/Card.module.css';
  composes: cardHeading;
  height: auto;
  background-position: right center;
  text-decoration: none;
}

.addCollapse {
  composes: expand;
  background-image: url('../../components/images/chevron-up.svg');
}

.collapsedCard {
  composes: collapsedCard from '../../components/Card/Card.module.css';
  padding-left: 0;
}

.collapsedUrl {
  composes: collapsedName from '../../components/Card/Card.module.css';
  width: 50rem;
}

.collapsedType {
  composes: collapsedValue from '../../components/Card/Card.module.css';
}

.buttons {
  position: absolute;
  top: 0.75rem;
  right: 1.5rem;
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 0.75rem;

  --button-size: 0.75rem;
}

.collapse {
  composes: collapse from '../../components/Card/Card.module.css';
}

.remove {
  composes: buttonIcon from '../../components/Button/Button.module.css';
  width: var(--button-size);
  height: var(--button-size);
  background-image: url('../../components/images/trash-bin.svg');
}

.labelLine {
  position: relative;
}

.label {
  composes: smallText from '../../components/Typography/Typography.module.css';
  font-size: 0.625rem;
  display: block;
  width: 100%;
  margin-block: 0.5rem;
}

.input {
  composes: input from '../../components/Input/Input.module.css';
  width: calc(100% - 2rem);
  height: 1.5rem;
}

.buttonsLine {
  composes: buttonsLine from '../../components/Button/Button.module.css';
  margin-block: 1.5rem;
}

.submit {
  composes: buttonPrimary from '../../components/Button/Button.module.css';
}

.cancel {
  composes: buttonSecondary from '../../components/Button/Button.module.css';
}

.errorTooltipField {
  composes: tooltip from '../../components/ErrorTooltip/ErrorTooltip.module.css';
  top: 2.5rem;
  left: 60%;
}

.errorTooltip {
  composes: tooltip from '../../components/ErrorTooltip/ErrorTooltip.module.css';
  top: 2.5rem;
  left: 70%;
}

.endpoint {
  composes: card;
}

.name {
  composes: detailName from '../../components/Card/Card.module.css';
  margin-top: 0.5rem;
}

.value {
  composes: detailValue from '../../components/Card/Card.module.css';
  margin-top: 0.125rem;
  margin-bottom: 0.75rem;
  word-break: break-word;
}
