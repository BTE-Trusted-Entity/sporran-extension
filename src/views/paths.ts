export { generatePath } from 'react-router-dom';

export const paths = {
  home: '/',
  settings: '/settings',
  access: '/access',
  identity: {
    base: '/identity/',
    add: '/identity/add',
    create: {
      start: '/identity/create',
      backup: '/identity/create/backup',
      verify: '/identity/create/verify',
      password: '/identity/create/password',
    },
    import: {
      start: '/identity/import',
      password: '/identity/import/password',
    },

    overview: '/identity/:address/:type(created|imported|pwreset)?',
    reset: {
      start: '/identity/:address/reset',
      password: '/identity/:address/reset/password',
    },
    send: {
      start: '/identity/:address/send',
      review: '/identity/:address/send/review',
      warning: '/identity/:address/send/warning',
    },
    receive: '/identity/:address/receive',
    credentials: {
      base: '/identity/:address/credentials',
      presentation: '/identity/:address/credentials/:hash/presentation',
    },
    remove: '/identity/:address/remove',
    did: {
      upgrade: {
        start: '/identity/:address/did/upgrade',
        kilt: '/identity/:address/did/upgrade/kilt',
        euro: '/identity/:address/did/upgrade/euro',
      },
      manage: {
        start: '/identity/:address/did',
        downgrade: {
          base: '/identity/:address/did/downgrade',
          warning: {
            web3name: '/identity/:address/did/downgrade/warning/web3name',
            credentials: '/identity/:address/did/downgrade/warning/credentials',
          },
          sign: '/identity/:address/did/downgrade/sign',
        },
        endpoints: {
          start: '/identity/:address/did/endpoints',
          edit: '/identity/:address/did/endpoints/:id?',
          sign: '/identity/:address/did/endpoints/sign',
        },
      },
      repair: '/identity/:address/did/repair',
    },
    web3name: {
      create: {
        base: '/identity/:address/web3name/create',
        info: '/identity/:address/web3name/create/info',
        form: '/identity/:address/web3name/create/form',
        choose: '/identity/:address/web3name/create/choose/:web3name',
        kilt: '/identity/:address/web3name/create/kilt/:web3name',
        euro: '/identity/:address/web3name/create/euro/:web3name',
      },
      manage: {
        start: '/identity/:address/web3name',
        remove: '/identity/:address/web3name/remove',
      },
    },
  },
  popup: {
    base: '/popup',
    import: '/popup/import',
    access: '/popup/access',
    claim: '/identity/:address/claim',
    save: '/popup/save',
    share: {
      start: '/popup/share',
      sign: '/popup/share/sign',
    },
    sign: '/popup/sign',
    signRaw: '/popup/signRaw',
    signDid: {
      start: '/identity/:address/did/sign',
      sign: '/identity/:address/did/sign/password',
      credentials: '/identity/:address/did/sign/credentials',
    },
    signDidExtrinsic: '/identity/:address/did/signExtrinsic',
    createDid: '/identity/:address/did/create',
  },
};
