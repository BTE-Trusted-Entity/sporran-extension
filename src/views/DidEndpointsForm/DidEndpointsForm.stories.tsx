import { MemoryRouter, Route } from 'react-router-dom';
import { Meta } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import { identitiesMock } from '../../utilities/identities/IdentitiesProvider.mock';
import { generatePath, paths } from '../paths';

import { DidEndpointsForm } from './DidEndpointsForm';

export default {
  title: 'Views/DidEndpointsForm',
  component: DidEndpointsForm,
} as Meta;

const identity =
  identitiesMock['4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo'];

export function Template(): JSX.Element {
  return (
    <MemoryRouter
      initialEntries={[
        generatePath(paths.identity.did.manage.endpoints.start, {
          address: 'FOO',
        }),
      ]}
    >
      <Route path={paths.identity.did.manage.endpoints.edit}>
        <DidEndpointsForm
          identity={identity}
          onAdd={action('add')}
          onRemove={action('remove')}
        />
      </Route>
    </MemoryRouter>
  );
}
