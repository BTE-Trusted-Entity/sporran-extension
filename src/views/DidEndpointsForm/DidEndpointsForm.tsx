import {
  FormEvent,
  RefObject,
  useCallback,
  useEffect,
  useRef,
  useState,
} from 'react';
import { Link, Prompt, Redirect, useParams } from 'react-router-dom';
import { browser } from 'webextension-polyfill-ts';
import { stringToU8a } from '@polkadot/util';
import {
  ConfigService,
  Did,
  DidServiceEndpoint,
  KiltPublishedCredentialCollectionV1Type,
} from '@kiltprotocol/sdk-js';
import { last } from 'lodash-es';

import * as styles from './DidEndpointsForm.module.css';

import { IdentitySlide } from '../../components/IdentitySlide/IdentitySlide';
import { LinkBack } from '../../components/LinkBack/LinkBack';
import { Stats } from '../../components/Stats/Stats';
import { Identity } from '../../utilities/identities/types';
import { CopyValue } from '../../components/CopyValue/CopyValue';
import { useFullDidDocument } from '../../utilities/did/did';
import { useBooleanState } from '../../utilities/useBooleanState/useBooleanState';
import { generatePath, paths } from '../paths';
import { getIdentityDid } from '../../utilities/identities/identities';

function useScrollEndpoint(ref: RefObject<HTMLLIElement>, id: string) {
  const params: { id: string } = useParams();

  useEffect(() => {
    if (decodeURIComponent(params.id) === id && ref.current) {
      ref.current.scrollIntoView?.({ behavior: 'smooth', block: 'start' });
    }
  }, [id, params.id, ref]);
}

function DidEndpointCard({
  endpoint,
  startUrl,
  onRemove,
}: {
  endpoint: DidServiceEndpoint;
  startUrl?: string;
  onRemove: (endpoint: DidServiceEndpoint) => void;
}): JSX.Element {
  const t = browser.i18n.getMessage;

  const params: { id: string; address: string } = useParams();

  const handleDelete = useCallback(() => {
    onRemove(endpoint);
  }, [endpoint, onRemove]);

  const {
    type: [type],
    serviceEndpoint: [url],
  } = endpoint;
  const id = Did.resourceIdToChain(endpoint.id);

  const expanded = !startUrl || id === decodeURIComponent(params.id);
  const { address } = params;

  const ref = useRef<HTMLLIElement>(null);
  useScrollEndpoint(ref, endpoint.id);

  return (
    <li className={styles.endpoint} aria-expanded={expanded} ref={ref}>
      {startUrl && !expanded && (
        <Link
          className={styles.expand}
          to={generatePath(paths.identity.did.manage.endpoints.edit, {
            address,
            id,
          })}
          replace
        >
          <section className={styles.collapsedCard}>
            <h4 className={styles.collapsedUrl}>{url}</h4>
            <p className={styles.collapsedType}>{type}</p>
          </section>
        </Link>
      )}

      {expanded && (
        <section className={styles.buttons}>
          {startUrl && (
            <Link className={styles.collapse} to={startUrl} replace />
          )}
          <button
            type="button"
            className={styles.remove}
            onClick={handleDelete}
          />
        </section>
      )}

      {expanded && (
        <dl>
          <dt className={styles.name}>{t('view_DidEndpointsForm_url')}</dt>
          <dd className={styles.value}>{url}</dd>
          <dt className={styles.name}>{t('view_DidEndpointsForm_type')}</dt>
          <dd className={styles.value}>{type}</dd>
          <dt className={styles.name}>{t('view_DidEndpointsForm_id')}</dt>
          <dd className={styles.value}>{id}</dd>
        </dl>
      )}
    </li>
  );
}

function DidNewEndpoint({
  onAdd,
  tooMany,
  startUrl,
}: {
  onAdd: (endpoint: DidServiceEndpoint) => void;
  tooMany: boolean;
  startUrl?: string;
}): JSX.Element {
  const t = browser.i18n.getMessage;

  const params: { address: string; id: string } = useParams();
  const isAdding = params.id === 'add';

  const ref = useRef<HTMLLIElement>(null);
  useScrollEndpoint(ref, 'add');

  const dirty = useBooleanState();

  const namespace = ConfigService.get('api').consts.did;
  const maxIdLength = namespace.maxServiceIdLength.toNumber();
  const maxTypeLength = namespace.maxServiceTypeLength.toNumber();
  const maxUrlLength = namespace.maxServiceUrlLength.toNumber();

  const [endpointIdError, setEndpointIdError] = useState('');
  const [endpointUrlError, setEndpointUrlError] = useState('');
  const [endpointTypeError, setEndpointTypeError] = useState('');

  const handleSubmit = useCallback(
    async (event: FormEvent<HTMLFormElement>) => {
      event.preventDefault();

      setEndpointIdError('');
      setEndpointUrlError('');
      setEndpointTypeError('');

      if (tooMany) {
        return;
      }

      const formData = new FormData(event.currentTarget);
      const id = (formData.get('id') as string).trim();
      const url = (formData.get('url') as string).trim();
      const type = (formData.get('type') as string).trim();

      // allow all characters valid in the URI fragment except for %
      const unsupportedCharacter = id.match(/[^a-z0-9\-._~!$&'()*+=,;:@?/]/i);
      if (unsupportedCharacter) {
        setEndpointIdError(
          t('view_DidEndpointsForm_formatID', unsupportedCharacter),
        );
        return;
      }

      const idBytes = stringToU8a(id).length;
      if (maxIdLength && idBytes > maxIdLength) {
        setEndpointIdError(
          t('view_DidEndpointsForm_tooLong', [idBytes, maxIdLength]),
        );
        return;
      }

      const urlBytes = stringToU8a(url).length;
      if (maxUrlLength && urlBytes > maxUrlLength) {
        setEndpointUrlError(
          t('view_DidEndpointsForm_tooLong', [urlBytes, maxUrlLength]),
        );
        return;
      }

      const typeBytes = stringToU8a(type).length;
      if (maxTypeLength && typeBytes > maxTypeLength) {
        setEndpointTypeError(
          t('view_DidEndpointsForm_tooLong', [typeBytes, maxTypeLength]),
        );
        return;
      }

      dirty.off();
      // wait for React to update the dirty flag
      await new Promise((resolve) => setTimeout(resolve, 10));

      onAdd({
        id: `#${id}`,
        serviceEndpoint: [url],
        type: [type],
      });
    },
    [tooMany, maxIdLength, maxTypeLength, maxUrlLength, dirty, onAdd, t],
  );

  const isCollapsible = Boolean(startUrl);

  return (
    <li className={styles.add} aria-expanded={isAdding} ref={ref}>
      {!isCollapsible && (
        <h2 className={styles.cardHeading}>{t('view_DidEndpointsForm_add')}</h2>
      )}

      {isCollapsible && startUrl && (
        <Link
          className={isAdding ? styles.addCollapse : styles.expand}
          to={
            isAdding
              ? startUrl
              : generatePath(paths.identity.did.manage.endpoints.edit, {
                  address: params.address,
                  id: 'add',
                })
          }
          replace
        >
          {t('view_DidEndpointsForm_add')}
        </Link>
      )}

      {isAdding && (
        <form onSubmit={handleSubmit} onInput={dirty.on}>
          <Prompt
            when={dirty.current}
            message={t('view_DidEndpointsForm_unsaved')}
          />

          <div className={styles.labelLine}>
            <label className={styles.label}>
              {t('view_DidEndpointsForm_url')}
              <input className={styles.input} type="url" name="url" required />
            </label>
            {endpointUrlError && (
              <output className={styles.errorTooltipField}>
                {endpointUrlError}
              </output>
            )}
          </div>

          <div className={styles.labelLine}>
            <label className={styles.label}>
              {t('view_DidEndpointsForm_type')}
              <datalist id="types">
                <option value={KiltPublishedCredentialCollectionV1Type} />
              </datalist>
              <input
                className={styles.input}
                name="type"
                list="types"
                required
              />
            </label>
            {endpointTypeError && (
              <output className={styles.errorTooltipField}>
                {endpointTypeError}
              </output>
            )}
          </div>

          <div className={styles.labelLine}>
            <label className={styles.label}>
              {t('view_DidEndpointsForm_id')}
              <input
                className={styles.input}
                name="id"
                required
                defaultValue={String(Math.random()).substring(2, 8)}
              />
              {endpointIdError && (
                <output className={styles.errorTooltipField}>
                  {endpointIdError}
                </output>
              )}
            </label>
          </div>

          <p className={styles.buttonsLine}>
            {startUrl && (
              <Link className={styles.cancel} to={startUrl} replace>
                {t('common_action_cancel')}
              </Link>
            )}
            <button type="submit" className={styles.submit} disabled={tooMany}>
              {t('common_action_next')}
            </button>
            {tooMany && (
              <output className={styles.errorTooltip}>
                {t('view_DidEndpointsForm_tooMany')}
              </output>
            )}
          </p>
        </form>
      )}
    </li>
  );
}

interface Props {
  identity: Identity;
  onAdd: (endpoint: DidServiceEndpoint) => void;
  onRemove: (endpoint: DidServiceEndpoint) => void;
}

export function DidEndpointsForm({
  identity,
  onAdd,
  onRemove,
}: Props): JSX.Element {
  const t = browser.i18n.getMessage;

  const { address } = identity;
  const did = getIdentityDid(identity);

  const document = useFullDidDocument(did);

  const params: { id: string } = useParams();

  if (!document) {
    return <div className={styles.loading} />;
  }

  const endpoints = document.service || [];

  const api = ConfigService.get('api');
  const maxEndpoints = api.consts.did.maxNumberOfServicesPerDid.toNumber();

  const lastEndpoint = last(endpoints);
  const hasTooManyEndpoints = Boolean(endpoints.length >= maxEndpoints);
  const hasFewEndpoints = endpoints.length < 7;
  const hasNoEndpoints = endpoints.length === 0;
  const collapsible = !hasNoEndpoints;

  if (hasNoEndpoints && params.id !== 'add') {
    return (
      <Redirect
        to={generatePath(paths.identity.did.manage.endpoints.edit, {
          address,
          id: 'add',
        })}
      />
    );
  }

  const startUrl = generatePath(paths.identity.did.manage.endpoints.start, {
    address,
  });

  return (
    <section className={styles.container}>
      <h1 className={styles.heading}>{t('view_DidEndpointsForm_heading')}</h1>
      <p className={styles.subline}>{t('view_DidEndpointsForm_subline')}</p>

      <IdentitySlide identity={identity} />

      <CopyValue value={did} label="DID" className={styles.didLine} />

      <ul className={styles.list}>
        <DidNewEndpoint
          onAdd={onAdd}
          tooMany={hasTooManyEndpoints}
          startUrl={collapsible ? startUrl : undefined}
        />

        {endpoints.map((endpoint) => (
          <DidEndpointCard
            key={endpoint.id}
            endpoint={endpoint}
            startUrl={
              hasFewEndpoints && endpoint === lastEndpoint
                ? undefined
                : startUrl
            }
            onRemove={onRemove}
          />
        ))}
      </ul>

      {/* Use the link form because we could have been redirected to this form after submitting the transaction,
          and then going back from manage DID would end in a wrong place. */}
      <LinkBack to={paths.identity.did.manage.start} />
      <Stats />
    </section>
  );
}
