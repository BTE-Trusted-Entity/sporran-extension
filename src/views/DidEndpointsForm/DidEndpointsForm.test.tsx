import { MemoryRouter, Route } from 'react-router-dom';
import {
  ConfigService,
  connect,
  DidDocument,
  DidServiceEndpoint,
} from '@kiltprotocol/sdk-js';

import { identitiesMock, render, screen } from '../../testing/testing';
import { useFullDidDocument } from '../../utilities/did/did';
import { generatePath, paths } from '../paths';
import '../../components/useCopyButton/useCopyButton.mock';

import { DidEndpointsForm } from './DidEndpointsForm';

const identity =
  identitiesMock['4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo'];

const service: DidServiceEndpoint[] = [
  {
    serviceEndpoint: ['https://sporran.org/'],
    type: ['Some Type'],
    id: '#123456',
  },
  {
    serviceEndpoint: ['https://kilt.io/'],
    type: ['Another Type'],
    id: '#654321',
  },
];

const api = {
  consts: {
    did: {
      maxServiceIdLength: { toNumber: () => 50 },
      maxServiceTypeLength: { toNumber: () => 50 },
      maxServiceUrlLength: { toNumber: () => 200 },
      maxNumberOfServicesPerDid: { toNumber: () => 25 },
    },
  },
} as unknown as Awaited<ReturnType<typeof connect>>;
ConfigService.set({ api });

describe('DidEndpointsForm', () => {
  it('should match the snapshot', async () => {
    jest.mocked(useFullDidDocument).mockReturnValue({ service } as DidDocument);

    const mock = jest.spyOn(Math, 'random').mockReturnValue(0.123456);
    const { container } = render(
      <MemoryRouter
        initialEntries={[
          generatePath(paths.identity.did.manage.endpoints.start, {
            address: 'FOO',
          }),
        ]}
      >
        <Route path={paths.identity.did.manage.endpoints.edit}>
          <DidEndpointsForm
            identity={identity}
            onAdd={jest.fn()}
            onRemove={jest.fn()}
          />
        </Route>
      </MemoryRouter>,
    );

    expect(container).toMatchSnapshot();
    mock.mockClear();
  });

  it('should match the snapshot when no service endpoints', async () => {
    jest.mocked(useFullDidDocument).mockReturnValue({} as DidDocument);

    const mock = jest.spyOn(Math, 'random').mockReturnValue(0.123456);
    const { container } = render(
      <MemoryRouter
        initialEntries={[
          generatePath(paths.identity.did.manage.endpoints.start, {
            address: 'FOO',
          }),
        ]}
      >
        <Route path={paths.identity.did.manage.endpoints.edit}>
          <DidEndpointsForm
            identity={identity}
            onAdd={jest.fn()}
            onRemove={jest.fn()}
          />
        </Route>
      </MemoryRouter>,
    );

    await screen.getByLabelText('URL');

    expect(container).toMatchSnapshot();
    mock.mockClear();
  });
});
