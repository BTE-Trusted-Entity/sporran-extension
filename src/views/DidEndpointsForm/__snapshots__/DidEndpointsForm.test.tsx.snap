// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DidEndpointsForm should match the snapshot 1`] = `
<div>
  <section
    class="container"
  >
    <h1
      class="heading"
    >
      Service Endpoints
    </h1>
    <p
      class="subline"
    >
      Add and remove service endpoints
    </p>
    <section
      class="container"
    >
      <div
        class="avatar"
      >
        Identicon
      </div>
      <div
        class="centeredNameLine"
      >
        <span
          class="name"
        >
          Identity with on-chain DID and lots of coins
        </span>
      </div>
    </section>
    <p
      class="didLine"
    >
      <input
        aria-label="DID"
        class="long"
        readonly=""
        value="did:kilt:4oeJ76hdj84xnwCNqijUHUCTmfwXgSZ4vmxLEiTEYgQdBCcZ"
      />
      <button
        aria-label="copy to clipboard"
        class="copy"
        title="copy to clipboard"
        type="button"
      />
    </p>
    <ul
      class="list"
    >
      <li
        aria-expanded="false"
        class="add"
      >
        <a
          class="expand"
          href="/identity/FOO/did/endpoints/add"
        >
          + Add service endpoint
        </a>
      </li>
      <li
        aria-expanded="false"
        class="endpoint"
      >
        <a
          class="expand"
          href="/identity/FOO/did/endpoints/123456"
        >
          <section
            class="collapsedCard"
          >
            <h4
              class="collapsedUrl"
            >
              https://sporran.org/
            </h4>
            <p
              class="collapsedType"
            >
              Some Type
            </p>
          </section>
        </a>
      </li>
      <li
        aria-expanded="true"
        class="endpoint"
      >
        <section
          class="buttons"
        >
          <button
            class="remove"
            type="button"
          />
        </section>
        <dl>
          <dt
            class="name"
          >
            URL
          </dt>
          <dd
            class="value"
          >
            https://kilt.io/
          </dd>
          <dt
            class="name"
          >
            Type
          </dt>
          <dd
            class="value"
          >
            Another Type
          </dd>
          <dt
            class="name"
          >
            Service ID
          </dt>
          <dd
            class="value"
          >
            654321
          </dd>
        </dl>
      </li>
    </ul>
    <a
      aria-label="Back"
      class="linkBack"
      href="/identity/FOO/did"
      title="Back"
    />
    <p
      class="stats"
    >
      You have 3 Identities with a
      <span
        class="balance"
      >
         Total Balance: 
        <span
          class="amount"
        >
          3.7020
           
          <span
            aria-label="KILT Coin"
            class="coin"
          />
        </span>
      </span>
    </p>
  </section>
</div>
`;

exports[`DidEndpointsForm should match the snapshot when no service endpoints 1`] = `
<div>
  <section
    class="container"
  >
    <h1
      class="heading"
    >
      Service Endpoints
    </h1>
    <p
      class="subline"
    >
      Add and remove service endpoints
    </p>
    <section
      class="container"
    >
      <div
        class="avatar"
      >
        Identicon
      </div>
      <div
        class="centeredNameLine"
      >
        <span
          class="name"
        >
          Identity with on-chain DID and lots of coins
        </span>
      </div>
    </section>
    <p
      class="didLine"
    >
      <input
        aria-label="DID"
        class="long"
        readonly=""
        value="did:kilt:4oeJ76hdj84xnwCNqijUHUCTmfwXgSZ4vmxLEiTEYgQdBCcZ"
      />
      <button
        aria-label="copy to clipboard"
        class="copy"
        title="copy to clipboard"
        type="button"
      />
    </p>
    <ul
      class="list"
    >
      <li
        aria-expanded="true"
        class="add"
      >
        <h2
          class="cardHeading"
        >
          + Add service endpoint
        </h2>
        <form>
          <div
            class="labelLine"
          >
            <label
              class="label"
            >
              URL
              <input
                class="input"
                name="url"
                required=""
                type="url"
              />
            </label>
          </div>
          <div
            class="labelLine"
          >
            <label
              class="label"
            >
              Type
              <datalist
                id="types"
              >
                <option
                  value="KiltPublishedCredentialCollectionV1"
                />
              </datalist>
              <input
                class="input"
                list="types"
                name="type"
                required=""
              />
            </label>
          </div>
          <div
            class="labelLine"
          >
            <label
              class="label"
            >
              Service ID
              <input
                class="input"
                name="id"
                required=""
                value="123456"
              />
            </label>
          </div>
          <p
            class="buttonsLine"
          >
            <button
              class="submit"
              type="submit"
            >
              Next Step
            </button>
          </p>
        </form>
      </li>
    </ul>
    <a
      aria-label="Back"
      class="linkBack"
      href="/identity/4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo/did"
      title="Back"
    />
    <p
      class="stats"
    >
      You have 3 Identities with a
      <span
        class="balance"
      >
         Total Balance: 
        <span
          class="amount"
        >
          3.7020
           
          <span
            aria-label="KILT Coin"
            class="coin"
          />
        </span>
      </span>
    </p>
  </section>
</div>
`;
